toolkit/tools/precacher/precacher.go

```diff
@@ -0,0 +1,409 @@
+		completed := downloaded + skipped + failed + unavailable
+		progressPercent := (float64(completed) / float64(total)) * 100
+		if progressPercent > lastProgressUpdate+progressIncrement || done {
+			logger.Log.Infof("Pre-caching: %3d%% ( downloaded: %4d, skipped: %4d, unavailable: %4d, failed: %4d )", int(progressPercent), downloaded, skipped, unavailable, failed)
+			lastProgressUpdate = progressPercent
+		}
+	}
+	return
+}
+
+// precachePackage will attempt to download the specified package. It will return a downloadResult struct via the results.
+// This function runs with best effort, so it will return a result (of type downloadResultTypeFailure) if any error occurs
+// rather than returning an error. The caller is expected to have added to the provided wait group, while this function is
+// responsible for removing itself from the wait group. As much processing as possible is done before acquiring the
+// network operations semaphore to minimize the time spent holding it.
+func precachePackage(pkg *repocloner.RepoPackage, availablePackages map[string]string, outDir string, wg *sync.WaitGroup, results chan<- downloadResult, netOpsSemaphore chan struct{}) {
+	const (
+		downloadRetryAttempts = 2
+		downloadRetryDuration = time.Second
+	)
+
+	defer wg.Done()
+
+	// File names are of the form "<name>-<version>.<distro>.<arch>.rpm"
+	pkgName := fmt.Sprintf("%s-%s.%s.%s", pkg.Name, pkg.Version, pkg.Distribution, pkg.Architecture)
+	fileName := fmt.Sprintf("%s.rpm", pkgName)
+	fullFilePath := path.Join(outDir, fileName)
+	result := downloadResult{
+		pkgName: fileName,
+	}
+
+	// Bail out early if the file already exists
+	if exists, _ := file.PathExists(fullFilePath); exists {
+		result.resultType = downloadResultTypeSkipped
+		results <- result
+		return
+	}
+
+	// Get the url for the package, or bail out if it is not available
+	url, ok := availablePackages[pkgName]
+	if !ok {
+		result.resultType = donwloadResultTypeUnavailable
+		results <- result
+		return
+	}
+
+	// Limit the number of concurrent network operations to avoid overloading the network. All work past this point
+	// is either network related, or trivial, so we can safely hold the semaphore until we are done.
+	netOpsSemaphore <- struct{}{}
+	defer func() {
+		<-netOpsSemaphore
+	}()
+
+	logger.Log.Debugf("Pre-caching '%s' from '%s'", fileName, url)
+	err := retry.Run(func() error {
+		err := network.DownloadFile(url, fullFilePath, nil, nil)
+		if err != nil {
+			logger.Log.Warnf("Pre-caching: Attempt to download (%s) failed. Error: %s", url, err)
+		}
+		return err
+	}, downloadRetryAttempts, downloadRetryDuration)
+
+	if err != nil {
+		logger.Log.Warnf("Pre-caching '%s' failed. Error: %s", fileName, err)
+		result.resultType = downloadResultTypeFailure
+		results <- result
+		return
+	} else {
+		logger.Log.Debugf("Pre-caching '%s' succeeded", fileName)
+		result.resultType = downloadResultTypeSuccess
+		results <- result
```

Comment:
Since this is all happening in multiple threads, I'm not sure, if we shouldn't move all of this logging to `monitorProgress` instead. Otherwise, in case of errors, the debug logs might end up very hard to read.

Reply:
Done, had most of it there anyways and I think I meant to move these bits as well.